File: Recursion/combinationSum2/main.cpp
#include <iostream>
#include <map>
#include <vector>

using namespace std;

// Use hashTable and unique vector
class Solution {
 public:
  vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
    map<int, int> hashTable;
    for (const auto& e : candidates) {
      hashTable[e]++;
    }

    // sort the vector and remove the duplicate elements
    sort(candidates.begin(), candidates.end());
    auto last = unique(candidates.begin(), candidates.end());
    candidates.erase(last, candidates.end());

    vector<vector<int>> ans;
    vector<int> temp;

    helper(hashTable, candidates, ans, temp, target, 0);

    return ans;
  }

  void helper(map<int, int>& hashTable, vector<int>& candidates,
              vector<vector<int>>& ans, vector<int>& temp, int target,
              int index) {
    if (target == 0) {
      ans.push_back(temp);
      return;
    }

    for (int i = index; i < candidates.size(); i++) {
      int num = candidates[i];

      if (target < num) break;
      if (hashTable[num] == 0) continue;

      temp.push_back(num);
      hashTable[num]--;
      helper(hashTable, candidates, ans, temp, target - num, i);
      temp.pop_back();
      hashTable[num]++;
    }
  }
};

// Without hashTable and unique vector
class Solution {
 public:
  vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
    vector<vector<int>> ans;
    vector<int> temp;
    sort(candidates.begin(), candidates.end());

    helper(candidates, ans, temp, target, 0);

    return ans;
  }

  void helper(vector<int>& candidates, vector<vector<int>>& ans,
              vector<int>& temp, int target, int index) {
    if (target == 0) {
      ans.push_back(temp);
      return;
    }

    for (int i = index; i < candidates.size(); i++) {
      int curNum = candidates[i];
      if (i > index && curNum == candidates[i - 1]) continue;
      if (target < curNum) break;

      temp.push_back(curNum);
      helper(candidates, ans, temp, target - curNum, i + 1);
      temp.pop_back();
    }
  }
};

File: Recursion/combinationSum2/go/main.go
package main

import "sort"

// Using Hast Table and unique candidates list
func CombinationSum2(candidates []int, target int) [][]int {
	var result [][]int
	frequencyMap := make(map[int]int)
	uniqueCandidates := []int{}

	// Count frequencies and build unique candidates list
	for _, c := range candidates {
		if frequencyMap[c] == 0 {
			uniqueCandidates = append(uniqueCandidates, c)
		}
		frequencyMap[c]++
	}

	backtrack(uniqueCandidates, target, 0, []int{}, &result, frequencyMap)
	return result
}

func backtrack(candidates []int, target, index int, current []int, result *[][]int, frequencyMap map[int]int) {
	if target == 0 {
		*result = append(*result, append([]int(nil), current...))
		return
	}

	if target < 0 {
		return
	}

	for i := index; i < len(candidates); i++ {
		candidate := candidates[i]

		if frequencyMap[candidate] == 0 {
			continue
		}

		frequencyMap[candidate]--
		current = append(current, candidate)
		backtrack(candidates, target-candidate, i, current, result, frequencyMap)
		current = current[:len(current)-1]
		frequencyMap[candidate]++
	}
}

// Using Sorting and skipping duplicates
func CombinationSum22(candidates []int, target int) [][]int {
	sort.Slice(candidates, func(i, j int) bool {
		return candidates[i] < candidates[j]
	})

	ans := [][]int{}
	helper2(candidates, target, 0, []int{}, &ans)
	return ans
}

func helper2(candidates []int, target int, index int, cur []int, ans *[][]int) {
	if target < 0 {
		return
	}

	if target == 0 {
		t := make([]int, len(cur))
		copy(t, cur)
		*ans = append(*ans, t)
		return
	}

	for i := index; i < len(candidates); i++ {
		if i > index && candidates[i] == candidates[i-1] {
			continue
		}

		cur = append(cur, candidates[i])
		helper2(candidates, target-candidates[i], i+1, cur, ans)
		cur = cur[:len(cur)-1]
	}
}
